pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java

```diff
@@ -260,6 +266,8 @@ protected void sendMessagesToConsumers(ReadType readType, List<Entry> entries) {
             // readMoreEntries should run regardless whether or not stuck is caused by
             // stuckConsumers for avoid stopping dispatch.
             readMoreEntries();
+        }  else if (currentThreadKeyNumber == 0) {
```

Comment:
Does this will introduce an infinite loop when the groupedEntries.size() > 0 but consumers under the subscription can't receive more messages due to the flow control?

Reply:
And there is a race condition when check currentThreadKeyNumber == 0 here since we are updating the keyNumbers in the async callback of the `consumer.sendMessages()`;

Reply:
When messagesForC is 0， consumer.sendMessages() will not be called,  So keyNumbers can not be update in the async callback of the consumer.sendMessages();   And keyNumbers.decrementAndGet() == 0 may be not true here.

I add logs in 'else if (currentThreadKeyNumber == 0) ', log is printed sometimes.

Reply:
Could you please check my first comment? 

> Does this will introduce an infinite loop when the groupedEntries.size() > 0 but consumers under the subscription can't receive more messages due to the flow control?

If we have keyNumbers=10 first, then dispatch to consumers but messagesForC=0 for each key, so the currentThreadKeyNumber will be 0 right? here will trigger readMoreEntries immediately.

Reply:
yes, when last step messagesForC=0 in once loop while groupedEntries.size() > 0，readMoreEntries will be trigger immediately， just in this one scene.  And if messagesForC=0  is not in the last step of one loop, it will not be trigger immediately.  And this change will not be stuck sending message to consumer.
This way just trigger readMoreEntries method  , and consumers receive more message may be block by flow control or permits.
>  If we have keyNumbers=10 first, then dispatch to consumers but messagesForC=0 for each key, so the currentThreadKeyNumber will be 0 right? here will trigger readMoreEntries immediately.

Reply:
But the problem is this will waste CPU cycles, if for the next loop, currentThreadKeyNumber still be 0, this will lead to trigger the next loop immediately, my suggestion is delay for a while to trigger the next loop if currentThreadKeyNumber==0 such as 100ms.